paddle/framework/threadpool.h

```diff
@@ -0,0 +1,137 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserve.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#pragma once
+#include <condition_variable>
+#include <cstdio>
+#include <functional>
+#include <iostream>
+#include <mutex>
+#include <queue>
+#include <thread>
+
+namespace paddle {
+namespace framework {
+
+class ThreadPool {
+ private:
+  typedef std::function<void()> Func;
+  int num_threads_;
+  int available_;
+  bool running_;
+  bool complete_;
+  std::queue<Func> tasks_;
+  std::vector<std::unique_ptr<std::thread>> threads_;
+  std::mutex mutex_;
+  std::condition_variable condition_;
```

Comment:
`condition_` doesn't mean anything, should have a proper name.

Reply:
`condition_variable` must use with a mutex as a pair, `mutex_` is used every where, and don't know which variable it's protecting.

Reply:
> condition_ doesn't mean anything, should have a proper name

> condition_variable must use with a mutex as a pair, mutex_ is used every where, and don't know which variable it's protecting

the mutex just protects the task queue, maybe we only need one mutex for now?